[addons][vfs] fix addon version update and not allow disable, uninstall and update in case of use

AlwinEsch opened the pull request:
## Description
Commit 1:
-------------------
**[addons][vfs] fix addon version update**

Before was by a VFS addon update still the old version shown and the new visible after Kodi restart.

This was coming with the binary addon instance handling and here where the new was created before the old was destroyed, in this case does the system create a new instance on already opened addon (which in this case the old version).

This change unload now before the VFS addon (if not used somewhere else) and create the new after them, with this the correct CAddonDll with new version becomes created.

Further is the way improved and only the changed addon restarted, all others stays untouched. Before was by update of one addon all VFS addons restarted.

Commit 2:
-------------------
**[addons][vfs] not allow disable, uninstall and update in case of use**

Before was there easily possible to uninstall the addon if it was in use.
There becomes a check added and the CVFSAddonCache asked about usage of addon, if yes the parts becomes prevented.

To make this, is a helper class added where register on CAddonDll. As the binary addons use instance handlers, becomes it hard to make another way.
A instance class is for started place, mostly independent from CAddon and as there can be not easy to maintain.

By VFS is a bit different instance use, there is always only one, where handle itself open of different parts at same time.
This checks now how much his shared_ptr is in use, if higher as 1 report that addon class outside used.

But the added CAddonDll parts can be also used for multiple instances of other types within addon, only need then his related manager arranged to them.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## User related:

- Prevent VFS addon changes (update, disable and uninstall) if in use
- Fix VFS addon problem, where Kodi has needed restart before to show new installed version

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
With vfs.rar and vfs.sacd, about vfs.rar it can takes longer, until addon allowed to change, as there can be much work,

<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):
Picture of Dialog without active playback
![Bildschirmfoto von 2020-08-15 03-32-33](https://user-images.githubusercontent.com/6879739/90302691-4e6f2400-dea8-11ea-8a47-bc7cb21106ed.png)

The same during playback via addon:
![Bildschirmfoto von 2020-08-15 03-33-14](https://user-images.githubusercontent.com/6879739/90302703-6181f400-dea8-11ea-8529-c75f0b052fc2.png)


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

phunkyfish commented:
Jenkins build this please

phunkyfish commented on xbmc/addons/VFSEntry.cpp:
```diff
@@ -25,7 +25,22 @@ CVFSAddonCache::~CVFSAddonCache()
   CServiceBroker::GetAddonMgr().Events().Subscribe(this, &CVFSAddonCache::OnEvent);
-  Update();
+
+  // Load all during Kodi start available VFS addons
```
`// Load all available VFS addons during Kodi start`

AlwinEsch replied:
Thanks for comments! Have changed the text about.

phunkyfish commented on xbmc/addons/binary-addons/AddonDll.h:
```diff
@@ -30,6 +30,18 @@ namespace ADDON
  */
 using ADDON_INSTANCE_HANDLER = const void*;
 
+/*!
+ * @brief Information class for use on addon type managers.
+ *
+ * This to ask via @ref CAddonDll the manager there that work can be performed.
+ * By multiple instances can it be harder to becomes informed about changes.
+ */
```
/*!
 * @brief Information class for use on addon type managers.
 *
 * This to query via @ref CAddonDll the manager so that work can be performed.
 * If there are multiple instances it be harder to be informed about any changes.
 */

phunkyfish reviewed (approved):
Just some minors in the comment wording,

AlwinEsch committed:
[addons][vfs] fix addon version update

Before was by a VFS addon update still the old version shown and the new
visible after Kodi restart.

This was coming with the binary addon instance handling and here where
the new was created before the old was destroyed, in this case does
the system create a new instance on already opened addon (which in this
case the old version).

This change unload now before the VFS addon (if not used somewhere else)
and create the new after them, with this the correct CAddonDll with new
version becomes created.

Further is the way improved and only the changed addon restarted, all
others stays untouched. Before was by update of one addon all VFS
addons restarted.

AlwinEsch committed:
[addons][vfs] not allow disable, uninstall and update in case of use

Before was there easily possible to uninstall the addon if it was in use.
There becomes a check added and the CVFSAddonCache asked about usage of
addon, if yes the parts becomes prevented.

To make this, is a helper class added where register on CAddonDll. As the
binary addons use instance handlers, becomes it hard to make another way.
A instance class is for started place, mostly independent from CAddon
and as there can be not easy to maintain.

By VFS is a bit different instance use, there is always only one, where
handle itself open of different parts at same time.
This checks now how much his shared_ptr is in use, if higher as 1 report
that addon class outside used.

But the added CAddonDll parts can be also used for multiple instances within
addon, only need then his related manager arranged to them.

AlwinEsch commented:
Jenkins build this please

phunkyfish reviewed (approved):
👍

ksooo commented:
Could it be that the same problem exists with updating PVR add-ons? I think Il have seen this issue some time ago.

AlwinEsch commented:
No for there only relate to vfs addons, can also look soon how is there on PVR.

As I know is his work partly equal and can be possible that it bring the same cases. Only can in future there a bit more as with new API prepared to allow create of several instances (to support several backends on same addon).

But should be possible to find ways for them too :+1:
